server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchSourceContext.java

```diff
@@ -84,12 +121,48 @@ public String[] excludes() {
         return this.excludes;
     }
 
+    public boolean includeVectors() {
+        return this.includeVectors;
+    }
+
     public boolean hasFilter() {
-        return this.includes.length > 0 || this.excludes.length > 0;
+        return this.includes.length > 0 || this.excludes.length > 0 || this.includeVectors == Boolean.FALSE;
     }
 
-    public SourceFilter filter() {
-        return new SourceFilter(includes, excludes);
+    public SourceFilter filter(MappingLookup mappingLookup) {
+        if (includeVectors()) {
+            return new SourceFilter(includes, excludes);
+        } else {
+            if (mappingLookup == null) {
+                throw new IllegalArgumentException("MappingLookup must not be null when filtering vectors");
```

Comment:
IMO this should be an `IllegalStateException` because it's a server-side error if the mapping lookup is not supplied in a code path where vectors are filtered.